src/IO/H5/Helpers.cpp

```diff
@@ -473,8 +473,15 @@ void write_to_attribute(const hid_t location_id, const std::string& name,
   const hid_t att_id = H5Acreate2(location_id, name.c_str(), h5_type<Type>(),
                                   space_id, h5p_default(), h5p_default());
   CHECK_H5(att_id, "Failed to create attribute '" << name << "'");
-  CHECK_H5(H5Awrite(att_id, h5_type<Type>(), static_cast<const void*>(&value)),
-           "Failed to write value: " << value);
+  if constexpr (std::is_same_v<Type, std::string>) {
+    const auto cstr = value.c_str();
+    CHECK_H5(H5Awrite(att_id, h5_type<Type>(), &cstr),
+             "Failed to write value: " << value);
```

Comment:
I'm very confused.  This seems to work, but every example I can find online passes this function a `char *`, while this passes a `char **`.  What's going on?

Reply:
I tried removing the `&` and got a seg fault

Reply:
Yes, this is the form that seems to work.  I'm just wondering if you know of any documentation that actually explains how to use the function or an explanation of why every example I can find is apparently wrong.  SpEC does the same thing as this, which makes me happier.

Reply:
No I don't know of any documentation for this. I just copied this from Nils Vu's patch he sent, and didn't question it.

Reply:
I can't find any explicit HDF5 documentation on this. A couple examples I found online actually don't write the attribute out because they use `H5S_NULL` instead of `H5S_SCALAR`. Below is an example code if you want to try digging into this. My best guess is that because the `H5S_SCALAR` is a "scalar string" that you must pass a pointer to the string, i.e. a pointer to the pointer of characters. Here's the `H5Awrite` documentation https://docs.hdfgroup.org/hdf5/develop/group___h5_a.html#title37 with the `void * buf` being what I interpret as the "pointer to scalar" which if the scalar is of type `char*` then I guess you'd want a `char**`. In the below example, if I switch from `H5S_SCALAR` to `H5S_NULL` the code runs but `h5dump ./Attributes.h5` shows the attribute isn't written. If I try `H5S_SIMPLE` everything explodes with cryptic HDF5 internals errors that I don't understand. So what I can tell you is that empirically, this is the right way to write a null-terminated variable-length string as an HDF5 attribute. It may be different for datasets, I don't remember. I do remember years ago when writing the helpers that getting strings working was a pain because online examples were essentially useless...

```cpp
 #include <hdf5.h>
 #include <string>

 #define H5FILE_NAME "Attributes.h5"
 #define ATTR_NAME "VarLenAttr" /* Name of the string attribute */

 #define DATASET_RANK  1   /* Rank and size of the dataset  */
 #define DATASET_SIZE  7

 int main() {
   hid_t file, dataset, attr;       /* File and dataset identifiers */
   hid_t dataset_space, attr_space; /* Dataset's and Attribute's dataspace id */
   hid_t attr_type;                 /* Attribute type */
   herr_t status;                   /* Return value */

   hsize_t fdim[] = {DATASET_SIZE};
   std::string data{"ASD"};

   file = H5Fcreate(H5FILE_NAME, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);

   dataset_space = H5Screate(H5S_SIMPLE);
   status = H5Sset_extent_simple(dataset_space, DATASET_RANK, fdim, NULL);

   dataset = H5Dcreate2(file, "my_dataset", H5T_NATIVE_INT, dataset_space,
                        H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

   attr_space = H5Screate(H5S_SCALAR);
   attr_type = H5Tcopy(H5T_C_S1);
   status = H5Tset_size(attr_type, H5T_VARIABLE);
   status = H5Tset_strpad(attr_type, H5T_STR_NULLTERM);
   status = H5Tset_cset(attr_type, H5T_CSET_ASCII);

   attr = H5Acreate2(dataset, ATTR_NAME, attr_type, attr_space, H5P_DEFAULT,
                     H5P_DEFAULT);

   const char* attr_string = data.c_str();
   status = H5Awrite(attr, attr_type, &attr_string);
   H5Aclose(attr);
   H5Tclose(attr_type);
   H5Sclose(attr_space);

   H5Fclose(file);
 }
```


Is this enough for you to be okay with this part of the diff or is there some explicit change request you still have on this part?

Reply:
Thanks for the code.  Yes, I don't think anything needs to change for this.  I just spent a lot of time confused about how the interface worked and hoped someone else knew better than me.

Reply:
You're welcome! Okay, sonuds good! I tried digging into the HDF5 internals but got stuck where they try to use structs with function pointer members as a way of mocking classes: https://github.com/HDFGroup/hdf5/blob/18321dee40b21e19c95bf0ef2f61e3ec71233fe9/src/H5VLconnector.h#L837 So I never got to the final "write the string to disk" stage since I have no idea where the "constructor" for this class is :(